D365fo 3.1.7 (#1932)
mulesoft/docs-connectors: microsoft-365-ops/3.1/modules/ROOT/pages/microsoft-365-finance-operations-connector-config-topics.adoc

= Microsoft Dynamics 365 for Finance and Operations Connector 3.1 Additional Configuration Information - Mule 4

In Microsoft Dynamics 365 for Finance and Operations Connector 3.1.7 and later, you can use custom logic to configure properties of complex types when creating entities.

== Configure Complex Types with Custom Logic

In Microsoft Dynamics 365 for Finance and Operations Connector 3.1.6 and earlier, suppose you use the following payload in the *Create* operation:

[source,json,linenums]
----
%dw 2.0
output application/java
---
{
    "Name": "GENERIC-CUSTOMER-NAME",
    "CustomerAccount": "CUSTOMER-ACCOUNT",
    "SalesCurrencyCode": "EUR",
    "CustomerGroupId": "GROUP-ID",
    "IsServiceDeliveryAddressBased": "No"
}
----

> `IsServiceDeliveryAddressBased` is a property of the `Microsoft.Dynamics.DataEntities.NoYes` type. `Microsoft.Dynamics.DataEntities.NoYes` is a complex object in the Microsoft data model and is an enum type. Using this payload in the *Create* operation throws an exception because the value for `IsServiceDeliveryAddressBased` is `No`, which is a string instead of an enum type.

Reviewer: ```suggestion
`IsServiceDeliveryAddressBased` is a property of the `Microsoft.Dynamics.DataEntities.NoYes` type, which is an enum type in the Microsoft data model. Using this payload in the *Create* operation throws an exception because the value for `IsServiceDeliveryAddressBased` is `No`, which is a string instead of an enum type.
```

---
{
    "Name": "GENERIC-CUSTOMER-NAME",
    "CustomerAccount": "CUSTOMER-ACCOUNT",
    "SalesCurrencyCode": "EUR",
    "CustomerGroupId": "GROUP-ID",
    "IsServiceDeliveryAddressBased": "No"
}
----

`IsServiceDeliveryAddressBased` is a property of the `Microsoft.Dynamics.DataEntities.NoYes` type. `Microsoft.Dynamics.DataEntities.NoYes` is a complex object in the Microsoft data model and is an enum type. Using this payload in the *Create* operation throws an exception because the value for `IsServiceDeliveryAddressBased` is `No`, which is a string instead of an enum type.

In Microsoft Dynamics 365 for Finance and Operations Connector 3.1.7 and later, use the newly added custom logic to properly configure the property and avoid exceptions:

[source,json,linenums]
----
%dw 2.0
output application/java
---
{
    "Name": "GENERIC-CUSTOMER-NAME",
    "CustomerAccount": "CUSTOMER-ACCOUNT",
    "SalesCurrencyCode": "EUR",
    "CustomerGroupId": "GROUP-ID",
    "IsServiceDeliveryAddressBased": {
            "muleAdditions": {
                "kind": "enum",
                "type": "Microsoft.Dynamics.DataEntities.NoYes"
            },
            "value" : "No"
        }
}
----

> The value for `IsServiceDeliveryAddressBased` must be an object type and must contain the following:

Reviewer: ```suggestion
The value for `IsServiceDeliveryAddressBased` must be an object type that contains the following fields:

> 

Reviewer: The bullet order of the fields should match what is shown in the code above the list.

* `value`
+
Key with the enum value as a string.

* `muleAdditions`
** `kind`
+

> Key that indicates the kind of complex object. Supported values are `enum`.

Reviewer: ```suggestion
Key that indicates the kind of complex object. The only supported value is `enum`.
```

** `type`
+
Key that indicates the type in the Microsoft data model. Supported values are any enum type in the Microsoft data model.
+

> There are two ways to find out the enum type:

Reviewer: ```suggestion
To find out the enum type:
```

> . In Microsoft Dynamics 365 for Finance and Operations Connector 3.1.7 and later, the improved metadata now shows complete naming of a data type.

Reviewer: According to the lead-in, these are two ways to do something and not two steps. They should have regular bullets instead of numbers.

> . In Microsoft Dynamics 365 for Finance and Operations Connector 3.1.7 and later, the improved metadata now shows complete naming of a data type.

Reviewer: ```suggestion
. In Microsoft Dynamics 365 for Finance and Operations Connector 3.1.7 and later, the improved metadata now shows the complete name of a data type.
```

+
image::enum.png[List of metadata that shows the complete naming of the data type]
+

> . Execute a GET request on \https://{tenant}.com/data/$metadata. The resulting XML file contains the entire data model.

Reviewer: ```suggestion
. Execute a GET request on `\https://{tenant}.com/data/$metadata`. The resulting XML file contains the entire data model.
```
